File: src/csharp/Product.cs
/******************************************************************************

@LICENCE@
 
 *****************************************************************************/

using System;
using System.Collections.Generic;
using System.Text;

namespace BuyAtAPI
{

    public class Product
    {

        private BuyAtAPIClient apiClient;
        private BuyatAffiliateEntitiesProductDetails soapProductDetails;
        private Programme programme;
        private Feed feed;

        /// <summary>
        /// Constructor, wrapping SOAP object
        /// </summary>
        public Product(BuyAtAPIClient apiClient, BuyatAffiliateEntitiesProduct soapProduct)
        {
            this.apiClient = apiClient;
            this.soapProductDetails = soapProduct.product;
        }

        /// <summary>
        /// Constructor, wrapping SOAP object
        /// </summary>
        public Product(BuyAtAPIClient apiClient, BuyatAffiliateEntitiesProductDetails soapProductDetails)
        {
            this.apiClient = apiClient;
            this.soapProductDetails = soapProductDetails;
        }

        /// <summary>
        /// The product ID
        /// </summary>
        public int ProductID
        {
            get
            {
                try
                {
                    return Convert.ToInt32(soapProductDetails.product_id);
                }
                catch (FormatException)
                {
                    return -1;
                }
            }
        }

        /// <summary>
        /// The product SKU
        /// </summary>
        public string ProductSKU
        {
            get
            {
                return soapProductDetails.product_sku;
            }
        }

        /// <summary>
        /// The product URL
        /// </summary>
        public string ProductURL
        {
            get
            {
                return soapProductDetails.product_url;
            }
        }

        /// <summary>
        /// The product name
        /// </summary>
        public string ProductName
        {
            get
            {
                return soapProductDetails.product_name;
            }
        }

        /// <summary>
        /// The brand name
        /// </summary>
        public string BrandName
        {
            get
            {
                return soapProductDetails.brand_name;
            }
        }

        /// <summary>
        /// The product description
        /// </summary>
        public string Description
        {
            get
            {
                return soapProductDetails.description;
            }
        }

        /// <summary>
        /// The online price of the product
        /// </summary>
        public float OnlinePrice
        {
            get
            {
                try
                {
                    return Convert.ToSingle(soapProductDetails.online_price);
                }
                catch (FormatException)
                {
                    return -1;
                }
            }
        }

        /// <summary>
        /// The 3-letter currency code for the product's currency
        /// </summary>
        public string Currency
        {
            get
            {
                return soapProductDetails.currency;
            }
        }

        /// <summary>
        /// The HTML-encoded currency symbol for the product's currency
        /// </summary>
        public string CurrencySymbol
        {
            get
            {
                return soapProductDetails.currency_symbol;
            }
        }

        /// <summary>
        /// The image URL
        /// </summary>
        public string ImageURL
        {
            get
            {
                return soapProductDetails.image_url;
            }
        }

        /// <summary>
        /// The ID of the product's programme
        /// </summary>
        public int ProgrammeID
        {
            get
            {
                try
                {
                    return Convert.ToInt32(soapProductDetails.programme_id);
                }
                catch (FormatException)
                {
                    return -1;
                }
            }
        }

        /// <summary>
        /// The name of the product's programme
        /// </summary>
        public string ProgrammeName
        {
            get
            {
                return soapProductDetails.programme_name;
            }
        }

        /// <summary>
        /// The URL of the product's programme
        /// </summary>
        public string getProgrammeURL()
        {
            return soapProductDetails.programme_url;
        }

        /// <summary>
        /// The product's level 1 category ID
        /// </summary>
        public int Level1CategoryID
        {
            get
            {
                try
                {
                    return Convert.ToInt32(soapProductDetails.level1_category_id);
                }
                catch (FormatException)
                {
                    return -1;
                }
            }
        }

        /// <summary>
        /// The product's level 1 category name
        /// </summary>
        public string Level1CategoryName
        {
            get
            {
                return soapProductDetails.level1_category_name;
            }
        }

        /// <summary>
        /// The product's level 2 category ID
        /// </summary>
        public int Level2CategoryID
        {
            get
            {
                try
                {
                    return Convert.ToInt32(soapProductDetails.level2_category_id);
                }
                catch (FormatException)
                {
                    return -1;
                }
            }
        }

        /// <summary>
        /// The product's level 2 category name
        /// </summary>
        public string Level2CategoryName
        {
            get
            {
                return soapProductDetails.level1_category_name;
            }
        }

        /// <summary>
        /// The product's feed ID
        /// </summary>
        public int FeedID
        {
            get
            {
                try
                {
                    return Convert.ToInt32(soapProductDetails.feed_id);
                }
                catch (FormatException)
                {
                    return -1;
                }
            }
        }

        /// <summary>
        /// The product's feed name
        /// </summary>
        public string FeedName
        {
            get
            {
                return soapProductDetails.feed_name;
            }
        }

        /// <summary>
        /// The programme associated with this product
        /// </summary>
        public Programme Programme
        {
            get
            {
                if (this.programme == null)
                {
                    this.programme = this.apiClient.GetProgrammeInfo(this.ProgrammeID);
                }
                return this.programme;
            }
        }

        /// <summary>
        /// The programme associated with this product
        /// </summary>
        public Feed Feed
        {
            get
            {
                if (this.feed == null)
                {
                    this.feed = this.apiClient.GetFeedInfo(this.FeedID);
                }
                return this.feed;
            }
        }
    }
}

File: src/java/at/buy/entities/Product.java
/******************************************************************************

@LICENCE@
 
 *****************************************************************************/

package at.buy.entities;

import at.buy.api.APIException;
import at.buy.api.BuyatAPIClient;

/**
 * Represents a product from a buy.at feed.
 *
 * @author buy.at
 * @version 1.0
 * @since 1.0
 */
public class Product extends at.buy.entities.Entity {

  /**
   * Product ID
   */
  private int productID;
  
  /**
   * SKU
   */
  private String productSKU;
  
  /**
   * Tracking deeplink URL.
   */
  private String productURL;

  /**
   * Product name.
   */
  private String productName;

  /**
   * Brand name.
   */
  private String brandName;

  /**
   * Product description.
   */
  private String description;

  /**
   * Price of the product.
   */
  private float onlinePrice;

  /**
   * Three-letter currency code.
   */
  private String currency;

  /**
   * HTML-safe currency symbol.
   */
  private String currencySymbol;

  /**
   * URL of the product image.
   */
  private String imageURL;

  /**
   * ID of programme this product belongs to
   */
  private int programmeID;
  
  /**
   * Name of programme this product belongs to
   */
  private String programmeName;

  /**
   * URL of programme this product belongs to
   */
  private String programmeURL;

  /**
   * ID of level 1 category this product belongs to
   */
  private int level1CategoryID;
  
  /**
   * Name of level 1 category this product belongs to
   */
  private String level1CategoryName;

  /**
   * ID of level 2 category this product belongs to
   */
  private int level2CategoryID;
  
  /**
   * Name of level 2 category this product belongs to
   */
  private String level2CategoryName;

  /**
   * ID of feed this product belongs to
   */
  private int feedID;
  
  /**
   * Name of feed this product belongs to
   */
  private String feedName;

  /**
   * The programme object this product belongs to
   */
  private Programme programme;
  
  /**
   * The feed object this product belongs to
   */
  private Feed feed;
  
  /**
   * The client used to create this product
   */
  private BuyatAPIClient client;
  
  /**
   * Create a product object
   * @param productID
   * @param productURL
   * @param productName
   * @param brandName
   * @param description
   * @param onlinePrice
   * @param currency
   * @param currencySymbol
   * @param imageURL
   * @param programmeID
   * @param programmeName
   * @param level1CategoryID
   * @param level1CategoryName
   * @param level2CategoryID
   * @param level2CategoryName
   * @param feedID
   * @param feedName
   * @param client
   */
  public Product(int productID, String productSKU, String productURL, String productName,
      String brandName, String description, float onlinePrice, String currency,
      String currencySymbol, String imageURL, int programmeID,
      String programmeName, String programmeURL, int level1CategoryID, 
      String level1CategoryName, int level2CategoryID, String level2CategoryName, 
      int feedID, String feedName, BuyatAPIClient client) {
    this.productID = productID;
    this.productSKU = productSKU;
    this.productURL = productURL;
    this.productName = productName;
    this.brandName = brandName;
    this.description = description;
    this.onlinePrice = onlinePrice;
    this.currency = currency;
    this.currencySymbol = currencySymbol;
    this.imageURL = imageURL;
    this.programmeID = programmeID;
    this.programmeName = programmeName;
    this.programmeURL = programmeURL;
    this.level1CategoryID = level1CategoryID;
    this.level1CategoryName = level1CategoryName;
    this.level2CategoryID = level2CategoryID;
    this.level2CategoryName = level2CategoryName;
    this.feedID = feedID;
    this.feedName = feedName;
    this.client = client;
  }

  /**
   * @return the productID
   */
  public int getProductID() {
    return productID;
  }

  /**
   * @return the productSKU
   */
  public String getProductSKU() {
    return productSKU;
  }

  /**
   * @return the productURL
   */
  public String getProductURL() {
    return productURL;
  }

  /**
   * @return the productName
   */
  public String getProductName() {
    return productName;
  }

  /**
   * @return the brandName
   */
  public String getBrandName() {
    return brandName;
  }

  /**
   * @return the description
   */
  public String getDescription() {
    return description;
  }

  /**
   * @return the onlinePrice
   */
  public float getOnlinePrice() {
    return onlinePrice;
  }

  /**
   * @return the currency
   */
  public String getCurrency() {
    return currency;
  }

  /**
   * @return the currencySymbol
   */
  public String getCurrencySymbol() {
    return currencySymbol;
  }

  /**
   * @return the imageURL
   */
  public String getImageURL() {
    return imageURL;
  }

  /**
   * @return the programmeID
   */
  public int getProgrammeID() {
    return programmeID;
  }

  /**
   * @return the programmeName
   */
  public String getProgrammeName() {
    return programmeName;
  }

  /**
   * @return the programmeURL
   */
  public String getProgrammeURL() {
    return programmeURL;
  }

  /**
   * @return the level1CategoryID
   */
  public int getLevel1CategoryID() {
    return level1CategoryID;
  }

  /**
   * @return the level1CategoryName
   */
  public String getLevel1CategoryName() {
    return level1CategoryName;
  }

  /**
   * @return the level2CategoryID
   */
  public int getLevel2CategoryID() {
    return level2CategoryID;
  }

  /**
   * @return the level2CategoryName
   */
  public String getLevel2CategoryName() {
    return level2CategoryName;
  }

  /**
   * @return the feedID
   */
  public int getFeedID() {
    return feedID;
  }

  /**
   * @return the feedName
   */
  public String getFeedName() {
    return feedName;
  }

  /**
   * Get the programme associated with this product.
   *
   * @return The programme associated with this product.
   * @throws APIException if the api call fails
   */
  public Programme getProgramme() throws APIException {
    if(null == this.programme) {
      this.programme = this.client.getProgrammeInfo(this.programmeID);
    }
    return this.programme;
  }

  /**
   * Get the feed associated with this product.
   *
   * @return The feed which contains this product.
   * @throws APIException if the api call fails
   */
  public Feed getFeed() throws APIException {
    if(null == this.feed) {
      this.feed = this.client.getFeedInfo(this.feedID);
    }
    return this.feed;
  }
}
